onnxruntime/test/testdata/custom_op_library/custom_op_library.cc

```diff
@@ -3,71 +3,156 @@
+
+    dml_op_desc.Type = DML_OPERATOR_ELEMENT_WISE_IDENTITY;
+    dml_op_desc.Desc = &dml_identity_op_desc;
   }
-};
 
-// legacy custom op registration with kernel creation and compute function that return an OrtStatusPtr
-struct CustomOpOne : Ort::CustomOpBase<CustomOpOne, KernelOne, true> {
-  OrtStatusPtr CreateKernelV2(const OrtApi& /* api */, const OrtKernelInfo* /* info */, void** op_kernel) const {
-    *op_kernel = reinterpret_cast<void*>(std::make_unique<KernelOne>().release());
-    return nullptr;
-  };
+  void Compute(OrtKernelContext& /*ctx*/,
+               const Ort::Custom::DmlContext& dml_ctx,
+               const Ort::Custom::Tensor<float>& /*input*/,
+               Ort::Custom::Tensor<float>& /*output*/) {
+    // step 1: get resources from dml context
+    auto* dml_device = dml_ctx.dml_device;
+    CUSTOM_ENFORCE(dml_device, "failed to get dml device");
 
-  const char* GetName() const { return "CustomOpOne"; };
+    auto* d3d12_device = dml_ctx.d3d12_device;
+    CUSTOM_ENFORCE(d3d12_device, "failed to get d3d12 device");
 
-#ifdef USE_CUDA
-  const char* GetExecutionProviderType() const { return "CUDAExecutionProvider"; };
-#endif
+    auto* cmd_list = dml_ctx.cmd_list;
+    CUSTOM_ENFORCE(cmd_list, "failed to get cmd list");
 
-  size_t GetInputTypeCount() const { return 2; };
-  ONNXTensorElementDataType GetInputType(size_t /*index*/) const { return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT; };
+    auto* cmd_recorder = dml_ctx.cmd_recorder;
+    CUSTOM_ENFORCE(cmd_recorder, "failed to get cmd recorder");
 
-  size_t GetOutputTypeCount() const { return 1; };
-  ONNXTensorElementDataType GetOutputType(size_t /*index*/) const { return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT; };
+    // todo - more logic here to finish the compute ...
+  }
+
+  UINT tensor_shape[8U] = {};
+  DML_TENSOR_DESC dml_tensor_desc{};
+  DML_BUFFER_TENSOR_DESC dml_buffer_tensor_desc = {};
+  DML_ELEMENT_WISE_IDENTITY_OPERATOR_DESC dml_identity_op_desc = {};
+  DML_OPERATOR_DESC dml_op_desc = {};
+  ComPtr<IDMLOperator> dml_op = {};
+  ComPtr<IDMLCompiledOperator> dml_compiled_op = {};
+  ComPtr<IDMLOperatorInitializer> dml_op_initializer = {};
+  DML_BINDING_TABLE_DESC dml_binding_table_desc = {};
+  ComPtr<IDMLBindingTable> dml_binding_table = {};
+  D3D12_DESCRIPTOR_HEAP_DESC desc_heap_desc = {};
+  ComPtr<ID3D12DescriptorHeap> d3d12_desc_heap = {};
+  // ComPtr<IDMLCommandRecorder> dml_cmd_recorder = {};
+  const OrtApi* api{};
 };
+#endif
+
+void KernelOneCpu(const Ort::Custom::Tensor<float>& X,
+                  const Ort::Custom::Tensor<float>& Y,
+                  Ort::Custom::Tensor<float>& Z) {
+  auto input_shape = X.Shape();
+  auto x_raw = X.Data();
+  auto y_raw = Y.Data();
+  auto z_raw = Z.Allocate(input_shape);
+  for (int64_t i = 0; i < Z.NumberOfElement(); ++i) {
+    z_raw[i] = x_raw[i] + y_raw[i];
+  }
+}
+
+#ifdef USE_CUDA
+void KernelOneCuda(const Ort::Custom::CudaContext& cuda_ctx,
```

Comment:
those #ifdef make the code here super messy. could we split this file into cpu_custom_op, cuda_custom_op and dml_custom_op, and selectively build them based on build option? so it give user a clear example about how to write custom op (with different EPs).

Reply:
Done, ops now placed to different files and folders, make files adjusted accordingly.